### PR TITLE
[typescript-axios] / #12828: Add ESM support to typescript-axios client

### PR DESCRIPTION
**Pull request by miniengineer**
fix https://github.com/OpenAPITools/openapi-generator/issues/12828

An additional tsconfig.esm.json is added to every generated typescript-axios code having supportsES6 and npmName.
That will create an additional ECMAScript Module (ESM) to the npm build.

I used similar PR for `typescript-fetch` client as reference https://github.com/OpenAPITools/openapi-generator/pull/11720

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov  @davidgamero @mkusaka

### CODE REVIEW

**Comment by miniengineer** [comment]
This is my first time contributing to open source 🙇‍♀️ 

About the following step ⬇️ 

>Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
./mvnw clean package 
./bin/generate-samples.sh
./bin/utils/export_docs_generators.sh

After running the first command there are some errors, but they are related to the Python client 🤔 
I run it while on latest master and I get the same errors, so I assume those are not related to changes in this PR.

When generating samples I also got an error T_T

```
./bin/generate-samples.sh ./bin/configs/typescript*
```

```
Error: Unable to access jarfile /blabla/openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar
```

I am sorry, Java is not my first language.
Am I doing something wrong?

**Review by jfeltesse-mdsol — APPROVED** [review]
mind blown cat

**Comment by miniengineer** [comment]
😍 😍 😍 😍

**Comment by miniengineer** [comment]
**UPDATE**

I was able to create the JAR file using the following command that skips tests
`mvn clean install -Dmaven.javadoc.skip=true -Dmaven.test.skip=true`

and successfully generated samples 👍

File: modules/openapi-generator/src/main/resources/typescript-axios/package.mustache [hunk 1]
```diff
@@ -17,8 +17,12 @@
   "license": "Unlicense",
   "main": "./dist/index.js",
   "typings": "./dist/index.d.ts",
+{{#supportsES6}}
+  "module": "./dist/esm/index.js",
+  "sideEffects": false,
+{{/supportsES6}}
   "scripts": {
-    "build": "tsc --outDir dist/",
+    "build": "tsc {{#supportsES6}}&& tsc -p tsconfig.esm.json{{/supportsES6}}",
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-axios/package.mustache:25]: please keep the existing behavior by adding `{{^supportsES6}}--outDir dist/{{/supportsES6}}`
```suggestion
    "build": "tsc {{#supportsES6}}&& tsc -p tsconfig.esm.json{{/supportsES6}}{{^supportsES6}}--outDir dist/{{/supportsES6}}",
```

Comment (jfeltesse-mdsol) [reply modules/openapi-generator/src/main/resources/typescript-axios/package.mustache:25]: isn't it already provided through https://github.com/OpenAPITools/openapi-generator/blob/94bc2ed19c75caf627a58f48d9cb022669292bb7/modules/openapi-generator/src/main/resources/typescript-axios/tsconfig.mustache#L7 though?

Comment (miniengineer) [reply modules/openapi-generator/src/main/resources/typescript-axios/package.mustache:25]: Thanks Julien

Yes, this is why I removed the `outDir`. We don't need it in the `package.json` file, as it's already mentioned in the `tsconfig`.

I already tested generating TS clients on my local 👍

**Comment by miniengineer** [comment]
@macjohnny 
Thank you for the review 🙏 

About removing `outDir`, as Julien mentioned (and I should've probably added that in the PR description) the reason why I deleted it is because we already supply output directory in the ts config file, so it's redundant in the `package.json`.

I have successfully build and tested out TS clients on my local 👍

File: modules/openapi-generator/src/main/resources/typescript-axios/package.mustache (hunk 1 above)
Comment (macjohnny) [reply modules/openapi-generator/src/main/resources/typescript-axios/package.mustache:25]: makes sense, thx

**Pull request merged**